main.tf

```diff
@@ -0,0 +1,361 @@
+  vpc_id            = aws_vpc.default.id
+  service_name      = "com.amazonaws.${var.region}.ssmmessages"
+  vpc_endpoint_type = "Interface"
+
+  security_group_ids = [
+    aws_security_group.vpc_endpoint.id
+  ]
+
+  subnet_ids          = aws_subnet.private.*.id
+  private_dns_enabled = true
+
+  tags = merge(
+    {
+      Name = "endpointSsmMessages"
+    },
+    var.tags
+  )
+}
+
+#
+# NAT resources
+#
+resource "aws_eip" "nat" {
+  count = length(var.public_subnet_cidr_blocks)
+
+  vpc = true
+
+  tags = merge(
+    {
+      Name = "ElasticIP"
+    },
+    var.tags
+  )
+}
+
+resource "aws_nat_gateway" "default" {
+  depends_on = [aws_internet_gateway.default]
+
+  count = length(var.public_subnet_cidr_blocks)
+
+  allocation_id = aws_eip.nat[count.index].id
+  subnet_id     = aws_subnet.public[count.index].id
+
+  tags = merge(
+    {
+      Name = "gwNAT"
+    },
+    var.tags
+  )
+}
+
+#
+# Bastion resources
+#
+resource "aws_security_group" "bastion" {
+  name_prefix = "sgBastion"
+  vpc_id      = aws_vpc.default.id
+
+  tags = merge(
+    {
+      Name = "sgBastion"
+    },
+    var.tags
+  )
+}
+
+resource "aws_security_group_rule" "bastion_vpc_endpoint_egress" {
+  type      = "egress"
+  from_port = 443
+  to_port   = 443
+  protocol  = "tcp"
+
+  security_group_id        = aws_security_group.bastion.id
+  source_security_group_id = aws_security_group.vpc_endpoint.id
+
+  description = "Allow outbound TCP traffic to the interface VPC endpoints on 443."
+}
+
+resource "aws_security_group_rule" "vpc_endpoint_bastion_ingress" {
+  type      = "ingress"
+  from_port = 443
+  to_port   = 443
+  protocol  = "tcp"
+
+  security_group_id        = aws_security_group.vpc_endpoint.id
+  source_security_group_id = aws_security_group.bastion.id
+
+  description = "Allow inbound TCP traffic from the bastion on 443."
+}
+
+data "aws_iam_policy_document" "ec2_assume_role" {
+  statement {
+    effect = "Allow"
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+
+    actions = ["sts:AssumeRole"]
+  }
+}
+
+resource "aws_iam_role" "bastion" {
+  name_prefix        = "BastionRole"
+  assume_role_policy = data.aws_iam_policy_document.ec2_assume_role.json
+
+  tags = var.tags
+}
+
+resource "aws_iam_role_policy_attachment" "ssm_managed_instance_core" {
+  role       = aws_iam_role.bastion.name
+  policy_arn = var.aws_ssm_managed_instance_core_policy_arn
+}
+
+resource "aws_iam_instance_profile" "bastion" {
+  name_prefix = "BastionInstanceProfile"
+  role        = aws_iam_role.bastion.name
+
+  tags = var.tags
+}
+
+resource "aws_instance" "bastion" {
+  ami                    = var.bastion_ami
+  availability_zone      = var.availability_zones[0]
+  ebs_optimized          = true
+  iam_instance_profile   = aws_iam_instance_profile.bastion.id
+  instance_type          = var.bastion_instance_type
+  monitoring             = true
+  subnet_id              = aws_subnet.private[0].id
```

Comment:
I like using SSM for accessing private EC2 instances, we actually use the same pattern currently with developer workspaces they can provision through the service catalog. It keeps in line with the Well-Architected framework, specifically [EC2_INSTANCE_MANAGED_BY_SSM](https://docs.aws.amazon.com/config/latest/developerguide/ec2-instance-managed-by-systems-manager.html)

That being said, I have two big problems I have with SSM, but I only think we can address one of those here.

1. I don't like connecting as `ssm-user` to the bastion host. If at all possible, I'd rather use a default user like `ubuntu` or `ec2-user` and disable the `ssm-user` altogether. You can use SSM as an SSH proxy-command instead. I've written some about how you can set it up that way, and this is pretty close to how we use it.
https://adamtheautomator.com/aws-bastion-host/

2. I think `AmazonSSMManagedInstanceCore` needs a conditional statement to limit it by something, either to a specific region or something else. If we wanted to use SSM for our build servers, then that's opening up another attack vector. This isn't something that can be fixed here, and we need to put some time into that to define those roles better, but that being said, it looks like we can pass an ARN in [here](https://github.com/d3b-center/terraform-aws-vpc/blob/9c9029ad615fda096504ccc2785fa65d1eabbacb/main.tf#L335) without adding too much complexity. Still, here's a good doc with some ways to add some conditionals to SSM that we can look at another time.
https://docs.aws.amazon.com/systems-manager/latest/userguide/security_iam_id-based-policy-examples.html

Reply:
Since we are building our own module, we should build in some level of best practices that we establish to be sufficient.  Using that IAM role is acceptable and it is ok to "recommend" to all projects. 
One thing to consider is that we would need an SSH tunnel to the VPC or some type of way to access services within private/public VPC. Similarly how we use "sshuttle" tool that is part of dev-env-tunnel is that we need to access multiple services at the same time(without ssh tunneling to localhost individual hosts), since they are used by portal UI.